qualtran/pyzx_interop/bloq_to_pyzx_circuit.py

```diff
@@ -0,0 +1,179 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Iterable
+
+import numpy as np
+import pyzx as zx
+from attrs import define
+from numpy.typing import NDArray
+
+from qualtran import Bloq, CompositeBloq, LeftDangle, Register, RightDangle, Signature, Soquet
+
+ZXQubitMap = dict[str, NDArray[np.integer]]
+"""A mapping from register names to an NDArray of ZX qubits"""
+
+
+@define
+class ZXAncillaManager:
+    """A simple ancilla qubit manager for generating pyzx Circuits.
+
+    Attributes:
+        n: number of existing qubits in the starting circuit.
+    """
+
+    n: int
+
+    def allocate(self) -> int:
+        """Allocate an uninitialized ancilla qubit.
+
+        This returns the index of a new ancilla qubit for use.
+        Note that it must be manually initialized, e.g. using the `InitAncilla` gate.
+        """
+        idx = self.n
+        self.n += 1
+        return idx
+
+    def free(self, q: int):
+        """Free an ancilla qubit.
+
+        Discard an ancilla qubit. For now, this operation does nothing.
+        """
+
+
+def _empty_qubit_map_from_registers(registers: Iterable[Register]) -> ZXQubitMap:
+    """For each register, creates an empty NDArray of the appropriate shape to store the zx qubits."""
+    return {reg.name: np.empty(reg.shape + (reg.bitsize,), dtype=int) for reg in registers}
+
+
+def _initalize_zx_circuit_from_signature(
+    signature: Signature,
+) -> tuple[zx.Circuit, ZXQubitMap, ZXAncillaManager]:
+    """Initalize a pyzx circuit from a bloq signature.
+
+    This enumerates the qubits in the same order as the registers in the signature.
+
+    Args:
+        signature: the signature of the bloq.
+
+    Returns:
+        A tuple of the pyzx circuit, the mapping from register names to qubits, and an
+        ancilla manager for the circuit.
+    """
+
+    n_qubits: int = 0
+    qubit_d: ZXQubitMap = {}
+
+    for reg in signature.lefts():
+        n = reg.total_bits()
+        idxs = np.arange(n) + n_qubits
+        shape = reg.shape + (reg.bitsize,)
+        qubit_d[reg.name] = idxs.reshape(shape)
+        n_qubits += n
+
+    circ = zx.Circuit(qubit_amount=n_qubits)
+    return circ, qubit_d, ZXAncillaManager(n_qubits)
+
+
+def _add_bloq_to_pyzx_circuit(
+    circ: zx.Circuit, bloq: Bloq, ancilla_manager: ZXAncillaManager, in_qubits: ZXQubitMap
+) -> ZXQubitMap:
+    """Add a single bloq acting on the given input qubits to a pyzx circuit.
+
+    Args:
+        circ: the pyzx circuit.
+        bloq: the bloq to add.
+        ancilla_manager: the ancilla manager for `circ`.
+        in_qubits: the input qubits to the bloq.
+
+    Returns:
+        A mapping of output register names to output qubits.
+    """
+    try:
+        gates, out_qubits = bloq.as_zx_gates(ancilla_manager, **in_qubits)
+        for gate in gates:
+            circ.add_gate(gate)
+        return out_qubits
+    except NotImplementedError:
+        pass
```

Comment:
This is so we fall back to the decomposition-based approach? Maybe some line comments could help guide the reader to the order of strategies